File: 3768.minimum-inversion-count-in-subarrays-of-fixed-length.py
#
# @lc app=leetcode id=3768 lang=python3
#
# [3768] Minimum Inversion Count in Subarrays of Fixed Length
#

# @lc code=start
class Solution:
    def minInversionCount(self, nums: List[int], k: int) -> int:
        def coordinate_compress(nums):
            unique_nums = sorted(set(nums))
            return {num: i + 1 for i, num in enumerate(unique_nums)}
        
        def add(fenwick_tree, index, value):
            while index < len(fenwick_tree):
                fenwick_tree[index] += value
                index += index & (-index)
        
        def sum(fenwick_tree, index):
            total = 0
            while index > 0:
                total += fenwick_tree[index]
                index -= index & (-index)
            return total
        
        def count_inversions(arr):
            inv_count = 0
            fenwick_tree = [0] * (len(arr) + 1)
            compressed_arr = [value_map[num] for num in arr]
            for i in reversed(range(len(compressed_arr))):
                inv_count += sum(fenwick_tree, compressed_arr[i] - 1)
                add(fenwick_tree, compressed_arr[i], 1)
            return inv_count
        
        n = len(nums)
        min_inversions = float('inf')
        value_map = coordinate_compress(nums)
        current_subarray = nums[:k]
        current_inv_count = count_inversions(current_subarray)
        min_inversions = min(min_inversions, current_inv_count)
         # Sliding window approach to update inversion count efficiently  fenwick_tree=[0]*(len(value_map)+1 )  compressed_arr=[value_map[num]for num in nums ] # Initialize the Fenwick Tree with the first k elements' inversions  for num in compressed_arr[:k]: add(fenwick_tree,num,+1 ) # Iterate over each sliding window position starting from the second window  for i in range(k,n ): # Remove the element going out of the window from Fenwick Tree influence  outgoing_value=compressed_arr[i-k ] add(fenwick_tree,outgoing_value,-1 ) current_inv_count-=sum(fenwick_tree,outgoing_value)-sum(fenwick_tree,outgoing_value-1 ) # Add new element coming into the window incoming_value=compressed_arr[i ] current_inv_count+=sum(fenwick_tree,len(value_map))-sum(fenwick_tree,incoming_value ) add(fenwick_tree,incoming_value,+1 ) min_inversions=min(min_inversions,current_inv_count ) return min_inversions # @lc code=end

File: 3768.minimum-inversion-count-in-subarrays-of-fixed-length.java
#
# @lc app=leetcode id=3768 lang=java
#
# [3768] Minimum Inversion Count in Subarrays of Fixed Length
#
# @lc code=start
import java.util.*;
class Solution {
    public long minInversionCount(int[] nums, int k) {
        int n = nums.length;
        // Discretize
        int[] sorted = nums.clone();
        Arrays.sort(sorted);
        Map<Integer, Integer> valToIdx = new HashMap<>();
        int idx = 1;
        for (int v : sorted) {
            if (!valToIdx.containsKey(v)) valToIdx.put(v, idx++);
        }
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) arr[i] = valToIdx.get(nums[i]);
        int size = idx + 2;
        BIT bit = new BIT(size);
        long inv = 0;
        // First window
        for (int i = 0; i < k; i++) {
            // Count how many greater elements before i
            inv += bit.query(size - 1) - bit.query(arr[i]);
            bit.update(arr[i], 1);
        }
        long minInv = inv;
        for (int i = k; i < n; i++) {
            // Remove arr[i - k]
            bit.update(arr[i - k], -1);
            inv -= bit.query(arr[i - k] - 1);
            // Add arr[i]
            inv += bit.query(size - 1) - bit.query(arr[i]);
            bit.update(arr[i], 1);
            minInv = Math.min(minInv, inv);
        }
        return minInv;
    }
    static class BIT {
        int[] tree;
        int n;
        BIT(int n) {
            this.n = n; tree = new int[n + 2];
        }
        void update(int x, int v) {
            while (x < tree.length) { tree[x] += v; x += x & -x; }
        }
        int query(int x) {
            int sum = 0; while (x > 0) { sum += tree[x]; x -= x & -x; } return sum;
        }
    }
}
# @lc code=end